mountpoint-s3-crt/src/s3/client.rs

```diff
@@ -545,11 +556,100 @@ impl Drop for MetaRequest {
+        let inner = unsafe { NonNull::new_unchecked(aws_s3_meta_request_acquire(self.inner.as_ptr())) };
+        Self { inner }
+    }
+}
+
 // SAFETY: `aws_s3_meta_request` is thread-safe
 unsafe impl Send for MetaRequest {}
 // SAFETY: `aws_s3_meta_request` is thread safe
 unsafe impl Sync for MetaRequest {}
 
+/// Future returned by `MetaRequest::write()`. It will complete when the write completes,
+/// or cancel the meta-request if cancelled.
+#[derive(Debug)]
+pub struct MetaRequestWrite {
+    receiver: oneshot::Receiver<Result<(), Error>>,
+    request: MetaRequest,
+}
+
+impl Drop for MetaRequestWrite {
+    fn drop(&mut self) {
+        if !self.receiver.is_terminated() {
+            // This future is being dropped before completion. Cancel the meta-request to ensure
+            // the client will not try to use the provided buffer.
+            self.request.cancel();
+        }
+    }
+}
+
+#[derive(Debug)]
+struct FutureVoid {
+    future: NonNull<aws_future_void>,
+    sender: oneshot::Sender<Result<(), Error>>,
+}
+
+// SAFETY: `aws_future_void` is thread-safe.
+unsafe impl Send for FutureVoid {}
+
+/// Safety: Don't call this function directly, only called by the CRT as a callback.
+unsafe extern "C" fn future_void_callback(user_data: *mut ::libc::c_void) {
+    let this = Box::from_raw(user_data as *mut FutureVoid);
+    let result = {
+        // SAFETY: `future` has completed.
+        let future_result = unsafe { aws_future_void_get_error(this.future.as_ptr()) };
+        let error_result: Error = future_result.into();
+        if error_result.is_err() {
+            Err(error_result)
+        } else {
+            Ok(())
+        }
+    };
+    let _ = this.sender.send(result);
+
+    // SAFETY: `self.future` contains a valid `aws_future_void`.
+    unsafe {
+        aws_future_void_release(this.future.as_ptr());
+    }
+}
+
+impl MetaRequestWrite {
+    fn new(meta_request: &MetaRequest, slice: &[u8], eof: bool) -> Self {
+        let request = meta_request.clone();
+
+        // SAFETY: `MetaRequestWrite` will ensure that `slice` is alive until the future completes or
+        // that the meta-request is cancelled if the future is cancelled, preventing further use of the
+        // `aws_byte_cursor`.
+        let data = unsafe { slice.as_aws_byte_cursor() };
+
+        // SAFETY: `aws_s3_meta_request_write` never returns NULL.
+        let future = unsafe { NonNull::new_unchecked(aws_s3_meta_request_write(request.inner.as_ptr(), data, eof)) };
+
+        let (sender, receiver) = oneshot::channel();
+        let sender = Box::leak(Box::new(FutureVoid { future, sender })) as *mut FutureVoid as *mut ::libc::c_void;
+
+        // SAFETY: `future` is a valid `aws_future_void` and this is the only callback we are registering.
+        unsafe {
+            aws_future_void_register_callback(future.as_ptr(), Some(future_void_callback), sender);
+        }
```

Comment:
I was a bit lost by calling the `FutureVoid` `sender`.

Does it sound good to name this `future`, `rust_future`, or something? `FutureVoid::future` could instead be `FutureVoid::inner`?